src/services/oss/dir_stream.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Result;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use bytes::Buf;
+use quick_xml::de;
+use serde::Deserialize;
+use time::format_description::well_known::Rfc3339;
+use time::OffsetDateTime;
+
+use super::backend::Backend;
+use super::error::parse_error;
+use crate::error::new_other_object_error;
+use crate::http_util::parse_error_response;
+use crate::object::ObjectPageStream;
+use crate::ops::Operation;
+use crate::path::build_rel_path;
+use crate::ObjectEntry;
+use crate::ObjectMetadata;
+use crate::ObjectMode;
+
+pub struct DirStream {
+    backend: Arc<Backend>,
+    root: String,
+    path: String,
+
+    token: String,
+
+    done: bool,
+}
+
+impl DirStream {
+    pub fn new(backend: Arc<Backend>, root: &str, path: &str) -> Self {
+        Self {
+            backend,
+            root: root.to_string(),
+            path: path.to_string(),
+
+            token: "".to_string(),
+
+            done: false,
+        }
+    }
+}
+
+#[async_trait]
+impl ObjectPageStream for DirStream {
+    async fn next_page(&mut self) -> Result<Option<Vec<ObjectEntry>>> {
+        if self.done {
+            return Ok(None);
+        }
+
+        let resp = self
+            .backend
+            .list_object(&self.path, self.token.clone())
+            .await?;
+
+        if resp.status() != http::StatusCode::OK {
+            if resp.status() == http::StatusCode::NOT_FOUND {
```

Comment:
Return error directly. List objects should never return 404.

Reply:
Unlike S3, OSS will return 404 on empty directories. 
When OpenDAL should return None.

Reply:
> Unlike S3, OSS will return 404 on empty directories, while OpenDAL should return None.

Incorrect, I'll ensure this.

Reply:
Negetive.

![image](https://user-images.githubusercontent.com/5351546/194487426-5f78561c-c5c2-4b81-8274-eb137df4bf61.png)